README.md

<!-- 
README.md is generated from README.Rmd. Please edit that file
rmarkdown::render("README.Rmd") 
-->
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/equidIO)](http://cran.r-project.org/package=equidIO) [![Travis-CI Build Status](https://travis-ci.org/bgctw/equidIO.svg?branch=master)](https://travis-ci.org/bgctw/equidIO)

Overview
--------

The `equidIO` package provides functionaly to deal with tabular data composed of records of equidistant time steps.

Main functions are

-   efficient reading of parts of such files: [readEquidistantCsv](https://github.com/bgctw/equidIO/tree/master/vignettes/readEquidistantCsv.md)
-   updating of grouped data.frames with ensuring equidistant time steps: [updateOutputRange](https://github.com/bgctw/equidIO/tree/master/vignettes/updateOutputRange.md)

Installation
------------

``` r
# From CRAN (in future)
#install.packages("equidIO")

# Or the the development version from GitHub:
# install.packages("devtools")
devtools::install_github("bgctw/equidIO")
```

Usage
-----

See the [package vignettes](https://github.com/bgctw/equidIO/tree/master/vignettes) (\*.md) for examples.


NEWS.md
# equidIO 0.0.3 

- remove dependeny on dplyrUtil
  harmonize only levels .group variable, others are handeled now by dplyr
- remove dependency on rlang

# equidIO 0.0.2

fix bug of passing down empty timezone argument to 'fCreateTimestamp'
in 'readEquidistantCsv'

# equidIO 0.0.1

basic setup


tests/testthat.R
library(testthat)
library(equidIO)

test_check("equidIO")


tests/testthat/test_readLoggerFile.R
.tmp.f <- function(){
  require(testthat)
  require(dplyr)
  require(rlang)
}
context("readEquidistantCsv")

testData <- dplyr::tribble(
  ~Date, ~Time, ~minute
  ,"2018-10-03", "06:00", 0
  ,"2018-10-03", "06:15", 15
  ,"2018-10-03", "06:30", 30
  ,"2018-10-03", "06:45", 45
  ,"2018-10-03", "07:00", 60
  ,"2018-10-03", "07:15", 75
) %>% mutate(minute = as.integer(.data$minute))

tmpDir <- tempdir()
#tmpDir <- "tmp"
tmpFile <- file.path(tmpDir, "testData.csv")
readr::write_csv(testData, tmpFile)
tmp <- readr::read_csv(tmpFile)

# generate csvFile with additional header lines not matching column types
lines <- readr::read_lines(tmpFile)
linesWithUnits <- c(
  lines[1]
  ,"-,-,min"
  ,"number,number,number"
  ,lines[-1]
)
tmpUnitsFile <- file.path(tmpDir, "testDataUnits.csv")
readr::write_lines(linesWithUnits, tmpUnitsFile)

linesWithHeader <- c(
  " Some Header information"
  , "across"
  , "three lines"
  , linesWithUnits
)
tmpHeaderFile <- file.path(tmpDir, "testDataHeader.csv")
readr::write_lines(linesWithHeader, tmpHeaderFile)

createTimestampDateTime <- function(
  ### add column timestamp from columns date and time
  data  ##<< data.frame to mutate
  , timezone = "GMT"
){
  data %>%  mutate(
    timestamp = as.POSIXct(
      paste(.data$Date, .data$Time), format = "%Y-%m-%d %H:%M", tz = timezone)
  )
}

test_that("readEquidistantCsv",{
  ans <- readEquidistantCsv(tmpFile, fCreateTimestamp = createTimestampDateTime )
  expect_equal( ans$minute, testData$minute)
  expect_equal( ans$timestamp[1], ISOdatetime(2018,10,3,6,0,0, tz = "GMT") )
})

test_that("readEquidistantCsv with timezone",{
  ans <- readEquidistantCsv(
    tmpFile, fCreateTimestamp = createTimestampDateTime, timezone = "UTC" )
  expect_equal( ans$minute, testData$minute)
  expect_equal( ans$timestamp[1], ISOdatetime(2018,10,3,6,0,0, tz = "UTC") )
})

testDataRe <- readEquidistantCsv(tmpFile, fCreateTimestamp = createTimestampDateTime )

test_that("readEquidistantCsv startTime",{
  startTime <- testDataRe$timestamp[3]
  ans <- readEquidistantCsv(
    tmpFile, fCreateTimestamp = createTimestampDateTime
    , startTime = startTime
    )
  expect_equal( ans$minute, testData$minute[-(1:2)])
})

test_that("readEquidistantCsv startTime and endTime",{
  startTime <- testDataRe$timestamp[3]
  endTime <- testDataRe$timestamp[5]
  ans <- readEquidistantCsv(
    tmpFile, fCreateTimestamp = createTimestampDateTime
    , startTime = startTime
    , endTime = endTime
  )
  expect_equal( ans$minute, testData$minute[3:5])
})

test_that("readEquidistantCsv startTime before file and endTime",{
  startTime <- testDataRe$timestamp[1] - 3600
  endTime <- testDataRe$timestamp[5]
  ans <- readEquidistantCsv(
    tmpFile, fCreateTimestamp = createTimestampDateTime
    , startTime = startTime
    , endTime = endTime
  )
  expect_equal( ans$minute, testData$minute[1:5])
})


test_that("readEquidistantCsv endTime",{
  endTime <- testDataRe$timestamp[5]
  ans <- readEquidistantCsv(
    tmpFile, fCreateTimestamp = createTimestampDateTime
    , endTime = endTime
  )
  expect_equal( ans$minute, testData$minute[1:5])
})


test_that("readEquidistantCsv endTime after file",{
  endTime <- testDataRe$timestamp[nrow(testDataRe)] + 3600
  ans <- readEquidistantCsv(
    tmpFile, fCreateTimestamp = createTimestampDateTime
    , endTime = endTime
  )
  expect_equal( ans$minute, testData$minute)
})

test_that("readEquidistantCsv error missing time steps",{
  testData2 <- testData %>% slice(-3)
  tmpFile2 <- file.path(tmpDir,"testData2.csv")
  readr::write_csv(testData2, tmpFile2)
  startTime <- testDataRe$timestamp[4] # after missing step
  expect_error(
    ans <- readEquidistantCsv(
      tmpFile2, fCreateTimestamp = createTimestampDateTime
      , startTime = startTime
    )
    ,"equidistant"

  )
})

test_that("readEquidistantCsv misaligned startTime",{
  startTime <- testDataRe$timestamp[3] + 60*7
  expect_error(
    ans <- readEquidistantCsv(
      tmpFile, fCreateTimestamp = createTimestampDateTime
      , startTime = startTime
    )
    ,"startTime must be a multiple"

  )
})

test_that("readEquidistantCsv misaligned endTime",{
  endTime <- testDataRe$timestamp[3] + 60*7
  expect_error(
    ans <- readEquidistantCsv(
      tmpFile, fCreateTimestamp = createTimestampDateTime
      , endTime = endTime
    )
    ,"multiple"
  )
})


test_that("readEquidistantCsv colHeader startTime and endTime",{
  startTime <- testDataRe$timestamp[3]
  endTime <- testDataRe$timestamp[5]
  ans <- readEquidistantCsv(
    tmpUnitsFile, nRowsColumnHeader = 2
    , fCreateTimestamp = createTimestampDateTime
    , startTime = startTime
    , endTime = endTime
  )
  expect_equal( ans$minute, testData$minute[3:5])
  dfHeader <- attr(ans, "columnHeader")
  expect_true("minute" %in% names(dfHeader))
  expect_equal( dfHeader$minute, c("min","number"))
})

test_that("readEquidistantCsv initial Header startTime and endTime",{
  startTime <- testDataRe$timestamp[3]
  endTime <- testDataRe$timestamp[5]
  ans <- readEquidistantCsv(
    tmpHeaderFile, nRowsColumnHeader = 2, nRowsHeader = 3
    , fCreateTimestamp = createTimestampDateTime
    , startTime = startTime
    , endTime = endTime
  )
  expect_equal( ans$minute, testData$minute[3:5])
  dfHeader <- attr(ans, "columnHeader")
  expect_true("minute" %in% names(dfHeader))
  expect_equal( dfHeader$minute, c("min","number"))
  headerLines <- attr(ans, "header")
  expect_equal( headerLines, linesWithHeader[1:3])
})

test_that("readEquidistantCsv initial Header startTime and endTime colTypes",{
  startTime <- testDataRe$timestamp[3]
  endTime <- testDataRe$timestamp[5]
  ans <- readEquidistantCsv(
    tmpHeaderFile, nRowsColumnHeader = 2, nRowsHeader = 3
    , fCreateTimestamp = createTimestampDateTime
    , startTime = startTime
    , endTime = endTime
    , col_types = readr::cols(
      Date = readr::col_character(), Time = readr::col_character())
  )
  expect_equal( ans$minute, testData$minute[3:5])
  dfHeader <- attr(ans, "columnHeader")
  expect_true("minute" %in% names(dfHeader))
  expect_equal( dfHeader$minute, c("min","number"))
  headerLines <- attr(ans, "header")
  expect_equal( headerLines, linesWithHeader[1:3])
})






tests/testthat/test_updateOutputRange.R
.tmp.f <- function(){
  library(testthat)
  library(dplyr)
  library(purrr)
  library(tidyr)
}

#require(testthat)
context("updateOutputRange")

dsTarget0 <- dplyr::tibble(
  date = seq(ISOdatetime(2010,1,1,0,0,30, tz = "UTC"), by = "30 min", length.out = 10)
  , resp = as.numeric(1:10)
  , sdResp = 10.0 + 1:10
  , canopyPosition = "treeCovered"
  , treatment = "Nadd"
)
dsTarget <- rbind( dsTarget0, dplyr::mutate(
  dsTarget0, canopyPosition = "openLand")) %>%
  dplyr::arrange(canopyPosition, date) %>%
  dplyr::mutate(canopyPosition = factor(canopyPosition))
indexColumns <- c("canopyPosition","treatment")

test_that("removeLastIncompleteRecord", {
  ds <- dsTarget %>% dplyr::group_by_at(
    dplyr::vars(dplyr::one_of(indexColumns)))
  ans <- removeLastIncompleteRecord(ds, "date")
  expect_equal( count(ans)$n, c(10,10))
  #
  ds <- dsTarget %>%
    mutate(
      date = as.POSIXct(
        ifelse(row_number() == n(), date - 10, date), origin = '1970-01-01')
    ) %>%
    group_by_at(vars(one_of(indexColumns)))
  ans <- removeLastIncompleteRecord(ds, "date")
  expect_equal( count(ans)$n, c(10,9))
})

test_that("updateOutputRange normal case",{
  dsNew <- dsTarget %>% slice(3:8) %>% mutate(resp = resp + 20.0)
  dsUp <- updateOutputRange(dsTarget, dsNew, indexColumns = indexColumns) %>%
    arrange(canopyPosition, date)
  expect_equal( nrow(dsUp), nrow(dsTarget))
  #problems with POSIXct: expect_equal( slice(dsUp, -(3:8)), slice(dsTarget, -(3:8)))
  expect_equal( slice(dsUp, -(3:8))$date, slice(dsTarget, -(3:8))$date)
  expect_equal( select(slice(dsUp, -(3:8)), -date),
                select(slice(dsTarget, -(3:8)), -date))
  #expect_equal( slice(dsUp, (3:8)), slice(dsNew))
  expect_equal( slice(dsUp, (3:8))$date, slice(dsNew)$date)
  expect_equal( select(slice(dsUp, (3:8)), -date), select(slice(dsNew), -date))
})

test_that("updateOutputRange additional times",{
  dsTarget1 <- slice(dsTarget, 1:18)
  dsNew <- slice(dsTarget, 17:20)
  iRowsT <- 1:16
  iRowsN <- 17:20
  dsUp <- updateOutputRange(dsTarget1, dsNew, indexColumns = indexColumns) %>%
    arrange(canopyPosition, date)
  expect_equal( nrow(dsUp), nrow(dsTarget))
  #problems with POSIXct: expect_equal( slice(dsUp, -(3:8)), slice(dsTarget, -(3:8)))
  expect_equal( slice(dsUp, iRowsT)$date, slice(dsTarget, iRowsT)$date)
  expect_equal( select(slice(dsUp, iRowsT), -date),
                select(slice(dsTarget, iRowsT), -date))
  #expect_equal( slice(dsUp, (3:8)), slice(dsNew))
  expect_equal( slice(dsUp, iRowsN)$date, slice(dsNew)$date)
  expect_equal( select(slice(dsUp, iRowsN), -date), select(slice(dsNew), -date))
})

test_that("updateOutputRange additional times with gap after target",{
  iRowsT <- 1:14
  iRowsN <- 17:20
  dsTarget1 <- slice(dsTarget, iRowsT)
  dsNew <- slice(dsTarget, iRowsN)
  dsUp <- updateOutputRange(dsTarget1, dsNew, indexColumns = indexColumns) %>%
    arrange(canopyPosition, date)
  expect_equal( nrow(dsUp), nrow(dsTarget))
  dsUp %>% group_by(across(indexColumns)) %>% nest() %>%
    mutate(ans = map_lgl(data, function(dss){
      expect_true( all(diff(as.numeric(dss$date)) == 1800))
    }))
  expect_equal( dsUp$resp[15], NA_real_ )
})

test_that("updateOutputRange additional times with gap before target",{
  iRowsT <- 4:20
  iRowsN <- 1:2
  dsTarget1 <- slice(dsTarget, iRowsT)
  dsNew <- slice(dsTarget, iRowsN)
  dsUp <- updateOutputRange(dsTarget1, dsNew, indexColumns = indexColumns) %>%
    arrange(canopyPosition, date)
  expect_equal( nrow(dsUp), nrow(dsTarget))
  dsUp %>% group_by(across(indexColumns)) %>% nest() %>%
    mutate(ans = map_lgl(data, function(dss){
      expect_true( all(diff(as.numeric(dss$date)) == 1800))
    }))
  expect_equal( dsUp$resp[3], NA_real_ )
})



test_that("updateOutputRange several groups",{
  iRowsN <- c(9,10,17:20)
  dsTarget1 <- slice(dsTarget, 1:18)
  dsNew <- slice(dsTarget, iRowsN)
  iRowsT <- setdiff(1:20, iRowsN)
  dsUp <- updateOutputRange(dsTarget1, dsNew, indexColumns = indexColumns) %>%
    arrange(canopyPosition, date)
  expect_equal( nrow(dsUp), nrow(dsTarget))
  #problems with POSIXct: expect_equal( slice(dsUp, -(3:8)), slice(dsTarget, -(3:8)))
  expect_equal( slice(dsUp, iRowsT)$date, slice(dsTarget, iRowsT)$date)
  expect_equal( select(slice(dsUp, iRowsT), -date),
                select(slice(dsTarget, iRowsT), -date))
  #expect_equal( slice(dsUp, (3:8)), slice(dsNew))
  expect_equal( slice(dsUp, iRowsN)$date, slice(dsNew)$date)
  expect_equal( select(slice(dsUp, iRowsN), -date), select(slice(dsNew), -date))
})

test_that("updateOutputRange additional factor level",{
  # If the new data.frame contains factors, with new levels. The target factor
  # is re-leveled with a warning.
  #  If the new data contains new levels of the index group, then records are added.
  iRowsN <- c(8:20)
  dsTarget1 <- slice(dsTarget, 1:10) %>% droplevels()
  expect_equal(length(levels(dsTarget1$canopyPosition)), 1L)
  dsNew <- slice(dsTarget, iRowsN) %>%  droplevels()
  iRowsT <- setdiff(1:20, iRowsN)
  # dplyrUitl issued a warning,  dplyr now can deal with unequal factor levels
  # but does not issue warnings
  #expect_warning(
    dsUp <- updateOutputRange(dsTarget1, dsNew, indexColumns = indexColumns) %>%
      arrange(canopyPosition, date)
  #)
  expect_equal( levels(dsUp), levels(dsTarget)) # added level in canopyPosition
  expect_equal( nrow(dsUp), nrow(dsTarget))
  #problems with POSIXct: expect_equal( slice(dsUp, -(3:8)), slice(dsTarget, -(3:8)))
  expect_equal( slice(dsUp, iRowsT)$date, slice(dsTarget, iRowsT)$date)
  expect_equal( select(slice(dsUp, iRowsT), -date),
                select(slice(dsTarget, iRowsT), -date))
  #expect_equal( slice(dsUp, (3:8)), slice(dsNew))
  expect_equal( slice(dsUp, iRowsN)$date, slice(dsNew)$date)
  expect_equal( select(slice(dsUp, iRowsN), -date), select(slice(dsNew), -date))
})

test_that("no index column",{
  # test if it also works if no index columns are present
  # use only the first 10 records of openLand
  iRowsN <- c(8:10)
  iRowsT <- 1:9
  iRowsTKept <- setdiff(iRowsT, iRowsN)
  dsTarget1 <- slice(dsTarget, iRowsT) %>%
    select(setdiff(names(dsTarget),indexColumns))
  dsNew <- slice(dsTarget, iRowsN) %>%
    select(setdiff(names(dsTarget),indexColumns))
  dsUp <- updateOutputRange(dsTarget1, dsNew) %>%
    arrange( date)
  expect_equal( nrow(dsUp), length(iRowsTKept) + length(iRowsN))
  # kept rows corresponds to target
  #problems with POSIXct: expect_equal( slice(dsUp, -(3:8)), slice(dsTarget, -(3:8)))
  expect_equal( slice(dsUp, iRowsTKept)$date, slice(dsTarget1, iRowsTKept)$date)
  expect_equal( select(slice(dsUp, iRowsTKept), -date)
                , select(slice(dsTarget1, iRowsTKept), -date))
  # new rows correspond to new
  #expect_equal( slice(dsUp, (3:8)), slice(dsNew))
  expect_equal( slice(dsUp, iRowsN)$date, slice(dsNew)$date)
  expect_equal( select(slice(dsUp, iRowsN), -date), select(slice(dsNew), -date))
})

test_that("bind distant timestamps",{
  skip("TODO")
})





R/donttest.R
donttest <- function(
  ### do not run the block of code non-interactively
  block                          ##<< block of code to be run or not
  ,run.donttest = interactive()  ##<< same as arguemnt in \code{\link{example}}.
    ## Logical scalar whether to evaluate the block of code.
    ## Defaults to TRUE in interactive sessions and FALSE otherwise.
){
  ##author<< Thomas Wutzler
  ##details<< Rd-files use the syntax \code{\\donttest{block}}
  ##, which is latex code (see \code{\link{example}}).
  ## In order to specify the same behaviour with valid R-code
  ##  this function can be used. This is necessary, if inlinedocs is used
  ##  to generate the Rd files.
  parentFrame = parent.frame()
  if (run.donttest) eval(block, envir = parentFrame) else NULL
  ##value<< return value of \code{eval(block)} or NULL if not evaluated.
}


R/dplyrUtil.R
left_joinReplace <- function(
  ### \code{\link{left_join}} but before drop columns that would be duplicated
  x,y   ##<< data.frames to join
  , by  ##<< here must be a character vector
  , ... ##<< further arguments to \code{\link{left_join}}
){
  ##details<< During joing existing columns are duplicated with a 
  ## different name.
  ## This function supports replacing the original columns instead,
  ## allowing for repeated join of similar data.
  # columns that will be created and need to be dropped from x before
  addCols <- setdiff(names(y),by)  
  delCols <- intersect(names(x), addCols)
  xD <- if (length(delCols)) select(x, -one_of(delCols)) else x
  ##value<< result of \code{\link{left_join}}
  left_join(xD, y, by = by, ...)
}

mapGroups <- function(
  ### split-map-combine
  data  ##<< groped data.frame
  , FUN  ##<< function(data.frmae, ...) -> data.frame to apply to subsets
  , ...  ##<< further arguments to FUN
  , drop = TRUE  ##<< logical indicating if levels that do not occur should 
  ## be dropped. Set to FALSE if FUN returns a data.frame also 
  ## for zero-row inputs.
){
  # https://coolbutuseless.bitbucket.io/2018/03/03/split-apply-combine-my-search-for-a-replacement-for-group_by---do/
  groupVars <- group_vars(data)
  if (!length(groupVars)) return(FUN(data,...))
  data %>% 
    split(select(.,groupVars), drop = drop) %>% 
    map_dfr(FUN,...)
}

mutate_cond <- function(
  ###  mutate acting only on the rows satisfying the condition
  .data           ##<< data.frame to be modified
  , condition     ##<< condition for selecting rows to be modified
  , ...           ##<< further arguments to \code{\link{mutate}}
  , envir = parent.frame()  ##<< the frame where the condition is evaluated
) {
  # credits to G. Grothendieck
  # https://stackoverflow.com/questions/34096162/dplyr-mutate-replace-on-a-subset-of-rows 
  ##details<< is not working on grouped data
  # TODO
  condition <- eval(substitute(condition), .data, envir)
  .data[condition, ] <- .data[condition, ] %>% mutate(...)
  .data
} 
attr(mutate_cond,"ex") <- function(){
  if (require(dplyr)) {
    ans <- iris %>% 
      mutate_cond(
        Species == "setosa"
        , Petal.Length = 1.0
      )
    ans %>% filter(Species == "setosa") %>% select(Petal.Length) %>% head() # 1.0
    ans %>% filter(Species != "setosa") %>% select(Petal.Length) %>% head() # orig
  }
}

#----------- expand factor levels ------------
# exported from equidIO

expandAllInconsistentFactorLevels <- function(
  ### expand a factor variables in all dataset to encompass levels of all sets
  ...  ##<< list of data.frames or several data.frames separated by comma
  , .noWarningCols = character(0)  ##<< string vector: do not warn for the these
  ## columns
) {
  dots <- list(...)
  datasets <- if (length(dots) == 1) dots[[1]] else dots
  colsToCheck <- intersect(names(datasets[[1]]),names(datasets[[2]]))
  # col <- colsToCheck[1]
  isInconsistentFactor <- sapply( colsToCheck,  function(col){
    any(map_lgl(
      datasets, ~is.factor(.[[col]]))) &&
      any(map_lgl(
        datasets, ~!identical(levels(.[[col]]), levels(datasets[[1]][[col]]))))
  })
  colNamesInc <- colsToCheck[isInconsistentFactor]
  colNamesWarn <- setdiff(colNamesInc, .noWarningCols)
  if (length(colNamesWarn)) warning(
    "releveling factors ", paste(colNamesWarn , collapse = ","))    
  for (col in colNamesInc) datasets <- expandFactorLevels(datasets, col)
  ##value<< \code{datasets} with updated factor columns
  datasets
}
attr(expandAllInconsistentFactorLevels,"ex") <- function(){
  if (exists("expandAllInconsistentFactorLevels")) {
    df1 <- data.frame(f = factor(c("D","D","C")))
    df2 <- data.frame(f = factor(c("C","C","A"))
                      , desc = c("forC1","forC2","forA1"))
    if (requireNamespace("dplyr"))
      dplyr::bind_rows(expandAllInconsistentFactorLevels(df1,df2))
    left_joinFactors(df1,df2)
  }
}

expandFactorLevels <- function(
  ### expand a factor in all dataset to encompass levels of all sets
  datasets    ##<< list of data.frames
  , varName   ##<< scalar string of variable holding the factor
) {
  #https://stackoverflow.com/questions/46876312/how-to-merge-factors-when-binding-two-dataframes-together/50503461#50503461
  groupLevels <- lvls_union(lapply(datasets, "[[", varName))
  force(varName)
  ans <- map(datasets, function(dss){
    #mutate(dss, !!varName := fct_expand(!!sym(varName), groupLevels))
    mutate(dss, !!varName := factor(!!sym(varName), levels = groupLevels))
  })
  ##value<< list of datasets with each entries column releveled
  ans
}

left_joinFactors <- function(
  ### left join with homogenizing factors before
  x     ##<< left tbl to join
  , y   ##<< right tbl to join
  , ... ##<< further arguments to \code{\link{left_join}} or \code{fJoin}
  , fJoin = left_join  ##<< join function(x,y,...) to use
){
  dfs <- expandAllInconsistentFactorLevels(x,y)
  ##value<< results of \code{\link{left_join}} or given alternative join function
  fJoin(dfs[[1]], dfs[[2]], ...)
}



R/zzzDebugCode.R
#+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
#+++ R script for deleting debugging code +++
#+++ Code used only for debugging can be written inside dummy functions .tmp.f().
#+++ With this script, these functions will be deleted on loading or sourcing
#+++ the package.
#+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

# Remove debugging code that is encapsulated in .tmp.f function blocks
# (to prevent R check complaints)
.tmp.f <- NULL

# Prevent R CMD check on complaining of unused bindings
if (getRversion() >= "2.15.1") {
	# used in magitr and classes
	utils::globalVariables(c(".", ".self",".x",".y","timestamp"))
}

# if (getRversion() >= "3.1.0") {
# 	utils::suppressForeignCheck(c("sID", "sDATA", "sINFO", "sLOCATION", "sTEMP", "sUSTAR")
# 	                              , package = "REddyProc")
# }
# 





R/updateOutputRange.R
#' @export
removeLastIncompleteRecord <- function(
  ### remove last row, if its timestep does not match (second - first row)
  data   ##<< the data.frame to check
  , colTimestamp = "timestamp"  ##<< scalar sting column name
  ## holding the time steps
){
  ##details<< Expects the timestamp column to hold end-of period timestamps
  ## with no missings.
  ## If the timestep is smaller than the first, this indicates that the
  ## data of the last time step was not complete yet.
  ## This method checks on this condition and removes an incomplete last row.
  fRemoveSingleGroup <- function(data, colTimestamp){
    if (nrow(data) <= 2) return(data)
    if (any(is.na(data[[colTimestamp]]))) stop(
      "no missings allowed in timestamp column ", colTimestamp)
    timestepFirst <- diff(head(data[[colTimestamp]], 2L))
    timestepLast <- diff(tail(data[[colTimestamp]], 2L))
    ##value<< \code{data} with last row removed, if the time step is incomplete
    if (timestepLast == timestepFirst) return(data)
    slice(data, -n())
  }
  .mapGroups(data, fRemoveSingleGroup, colTimestamp = colTimestamp)
}
attr(removeLastIncompleteRecord, "ex") <- function(){
  nRec <- 10
  dsComplete <- data.frame(date = seq(
    ISOdatetime(2010,1,1,0,0,30, tz = "UTC"), by = "30 min", length.out = nRec))
  dsIncomplete <- dsComplete
  dsIncomplete$date[nRec] <- dsIncomplete$date[nRec] - 2*60
  nrow( removeLastIncompleteRecord(dsComplete, "date") )
  nrow( removeLastIncompleteRecord(dsIncomplete, "date") )
}

# copied from lysiproc package
.mapGroups <- function(
  ### split-map-combine
  data  ##<< groped data.frame
  , FUN  ##<< function(data.frmae, ...) -> data.frame to apply to subsets
  , ...  ##<< further arguments to FUN
  , drop = TRUE  ##<< logical indicating if levels that do not occur should
  ## be dropped. Set to FALSE if FUN returns a data.frame also
  ## for zero-row inputs.
){
  # https://coolbutuseless.bitbucket.io/2018/03/03/split-apply-combine-my-search-for-a-replacement-for-group_by---do/
  groupVars <- group_vars(data)
  if (!length(groupVars)) return(FUN(data,...))
  data %>%
    split(select(.,groupVars), drop = drop) %>%
    map_dfr(FUN,...)
}

#--------- updateRData --------------
#' @export
updateRData <- function(
  ### update time-ordered data stored in RData file with new data
  newData         ##<< the updated data
  , fileName      ##<< scalar string path name of the RData file
  , objectName = file_path_sans_ext(basename(fileName)) ##<< scalar
  ## string: the name of the R-object. By default the basename of the
  , message = paste("updated", objectName, "in file", fileName) ##<< scalar
  ## string of a message to be displaced.
  , ...   ##<< further arguments to \code{\link{updateOutputRange}},
  ## such as \code{dateColumn} and \code{indexColumns}.
  , version = 2 ##<< RData version see \code{\link{save}}
){
  if (!file.exists(fileName)) {
    updatedData <- newData
  } else {
    ##details<< Assumes that fileName refers to an RData file with
    ## only one object inside
    origData <- local({load(fileName); get(ls()[1])})
    updatedData <- updateOutputRange(origData, newData, ...)
  }
  assign(objectName, updatedData)
  save( list = objectName, file = fileName, version = version)
  if (length(message))  base::message(message)
}

#' @export
updateOutputRange <- function(
  ### update part of the time-ordered data.frame with new values
  dsTarget              ##<< data.frame to update
  , dsNew               ##<< data.frame with new values
  , indexColumns = group_vars(dsNew)  ##<< other index columns beside date
  , dateColumn = "date" ##<< name of the column holding the dates/times
){
  ##details<< Update values of \code{dsNew} in \code{dsTarget}. Both
  ## data.frames must have the same columns. Each row is identified by
  ## data and values in the indexColumns.
  checkEqualColNames(dsTarget, dsNew)
  ##details<< The function requires that both data.frames have unique dates
  ## per index in equidistant time steps. The time steps must match.
  dsNew <- dsNew %>%
    create_group_column(indexColumns) %>%
    checkEquidistant(dateColumn, "Source")
  dsTarget <- dsTarget %>%
    create_group_column(indexColumns) %>%
    checkEquidistant(dateColumn, "Target")
  checkSameTimestep(dsTarget, dsNew, dateColumn)
  datasets <- expandFactorLevels(list(dsTarget, dsNew), ".group")
  dsTarget <- datasets[[1]]
  dsNew <- datasets[[2]]
  ##details<< Existing rows of the same index and time in the range of
  ## \code{dsNew}
  ## are dropped from \code{dsTarget}, and rows of \code{dsNew} are appended.
  newGroups <- unique(dsNew$.group)
  for (group in newGroups)
     dsTarget <- replaceGroup(dsTarget, dsNew, indexColumns, dateColumn, group)
  ##value<< argument \code{dsTarget} with rows of \code{dsNew} updated.
  ans <- dsTarget %>% ungroup() %>% select(-.data$.group)
  ans
}

checkEqualColNames <- function(dsTarget, dsNew) {
  iMissing <- which(is.na(match(names(dsNew), names(dsTarget))))
  if (length(iMissing)) stop(
    "The following columns do not exist in target: "
    , paste(names(dsNew)[iMissing], collapse = ","))
  iMissing <- which(is.na(match(names(dsTarget), names(dsNew))))
  if (length(iMissing)) stop(
    "The following columns do not exist in source: "
    , paste(names(dsTarget)[iMissing], collapse = ","))
}

create_group_column <- function(data, indexColumns) {
  nIndex <- length(indexColumns)
  data <- if (nIndex) {
    groupsVarsOrig <- group_vars(data)
    data %>%
      ungroup() %>%
      unite(".group", all_of(indexColumns), remove = FALSE) %>%
      mutate(.group = factor(.data$.group)) %>%
      group_by_at(vars(groupsVarsOrig))
  } else {
    data %>% mutate(.group = factor(1))
  }
  ##value<< \code{data} with new index column ".group"
  data
}

checkEquidistant <- function(data, dateColumn, dataName = "data") {
  data <- data %>% group_by(.data$.group)
  diffDate <- data %>%
    group_by(.data$.group) %>%
    arrange(.data[[dateColumn]]) %>%
    do(
      as.data.frame(table(diff(as.numeric(.[[dateColumn]]))))
    )
  if (any((diffDate %>%  count())$n != 1 )) stop(
    dataName, " has no equidistant time steps within groups")
  if (any( diffDate[[2]] != diffDate[[2]][1])) stop(
    dataName, "has no equidistant time across groups")
  ##value<< \code{data} grouped by index column ".group" and
  ## arranged ascending in dateColumn
  data
}

checkSameTimestep <- function(dsNew, dsTarget, dateColumn) {
  timestepNewSec <- diff(as.numeric(head(dsNew[[dateColumn]],2)))
  timestepTargetSec <- diff(as.numeric(head(dsTarget[[dateColumn]],2)))
  if (timestepNewSec != timestepTargetSec) stop(
    "Target has a different time step than source")
}

expandFactorLevels <- function(
  ### expand a factor in all dataset to encompass levels of all sets
  datasets    ##<< list of data.frames
  , varName   ##<< scalar string of variable holding the factor
){
  # copied from dplyrUtil
  #https://stackoverflow.com/questions/46876312/how-to-merge-factors-when-binding-two-dataframes-together/50503461#50503461
  groupLevels <- lvls_union(lapply(datasets, "[[", varName))
  force(varName)
  ans <- map(datasets, function(dss){
    dss[[varName]] <- factor(dss[[varName]], levels = groupLevels)
    dss
  })
  ##value<< list of datasets with each entries column releveled
  ans
}


replaceGroup <- function(
  ### replace a single group of new in target
  dsTarget, dsNew, indexColumns, dateColumn, group
) {
  dsNewGroup <- filter(dsNew, .data$.group == group)
  timestepSec <- diff(as.numeric(head(dsTarget[[dateColumn]],2)))
  dates <- dsNewGroup[[dateColumn]]
  # same group but before timestamp of new
  dsTargetGroupBefore <- dsTarget %>%
    filter(.data$.group == group) %>%
    filter(.data[[dateColumn]] < min(dates))
  # if new is after but not adjacent to target, need to fill empty lines
  dsFillBefore <- getFillBefore(
    dsTargetGroupBefore, dateColumn, min(dates), timestepSec, indexColumns)
  dsTargetGroupAfter <- dsTarget %>%
    filter(.data$.group == group) %>%
    filter(.data[[dateColumn]] > max(dates))
  # if fill is before target but not adjacent need to created fill lines
  dsFillAfter <- getFilledAfter(
    dsTargetGroupAfter, dateColumn, max(dates), timestepSec, indexColumns)
  dsTargetOtherGroups <- filter(dsTarget, .data$.group != group)
  dsTarget <- bind_rows(
    dsTargetOtherGroups
    , dsTargetGroupBefore
    , dsFillBefore
    , dsNewGroup
    , dsFillAfter
    , dsTargetGroupAfter
  )
}

getFillBefore <- function(
  ### create empty lines if new date is not adjacent after garget
  dsTargetGroupBefore, dateColumn, minDateNew, timestepSec, indexColumns
) {
  if (!nrow(dsTargetGroupBefore)) return(data.frame())
  maxBefore <- max(dsTargetGroupBefore[[dateColumn]])
  gapSeconds = as.numeric(minDateNew) - as.numeric(maxBefore)
  nDiffs = as.integer(gapSeconds / timestepSec)
  if (nDiffs != gapSeconds / timestepSec) stop(
    "times of target are misaligned with times of new")
  if (nDiffs == 1) return(data.frame())
  ##value<< data.frame with dateColumn and index columns or empty data.frame
  cbind.data.frame(setNames(data.frame(
    maxBefore + (1:(nDiffs - 1))*timestepSec
  ), dateColumn)
  , dsTargetGroupBefore[1,indexColumns,drop = FALSE])
}

getFilledAfter <- function(
  ### create empty lines if new date is not adjacent before garget
  dsTargetGroupAfter, dateColumn, maxDateNew, timestepSec, indexColumns
) {
  if (!nrow(dsTargetGroupAfter)) return(data.frame())
  minAfter <- min(dsTargetGroupAfter[[dateColumn]])
  gapSeconds = as.numeric(minAfter) - as.numeric(maxDateNew)
  nDiffs = as.integer(gapSeconds / timestepSec)
  if (nDiffs != gapSeconds / timestepSec) stop(
    "times of target are misaligned with times of new")
  if (nDiffs == 1) return(data.frame())
  ##value<< data.frame with dateColumn and index columns or empty data.frame
  cbind.data.frame(setNames(data.frame(
    maxDateNew + (1:(nDiffs - 1))*timestepSec
  ), dateColumn)
  , dsTargetGroupAfter[1,indexColumns,drop = FALSE])
}

.tmp.f <- function(){
  load("tmp/ETLys.RData")
  load("tmp/ETLysTmp.RData")
  ans <- updateOutputRange(ETLys, ETLysTmp, dateColumn = "timestamp")
}
